### PR TITLE
aarch64: First patch towards VcpuState save/restore

### PR DESCRIPTION
**Pull request by dianpopa**
## Reason for This PR

The current PR refactors arch/aarch64/src/regs.rs and adds changes for a smoother path towards saving and restoring vcpus on aarch64.

## Description of Changes

See #1826

- [ ] This functionality can be added in [`rust-vmm`](https://github.com/rust-vmm).

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The reason for this PR is clearly provided (issue no. or explanation).
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] Any newly added `unsafe` code is properly documented.
- [x] Any API changes are reflected in `firecracker/swagger.yaml`.
- [x] Any user-facing changes are mentioned in `CHANGELOG.md`.

### CODE REVIEW

File: src/arch/src/aarch64/regs.rs [hunk 1]
```diff
@@ -58,34 +65,35 @@ macro_rules! offset__of {
...
+        // in kernel by the kvm_regs structure. This structure is a
+        // mix of 32, 64 and 128 bit fields:
         // struct kvm_regs {
-        //	struct user_pt_regs regs;	/* sp = sp_el0 */
+        //      struct user_pt_regs regs;	/* sp = sp_el0 */
         //
-        //	__u64	sp_el1;
-        //	__u64	elr_el1;
+        //	    __u64	sp_el1;
+        //	    __u64	elr_el1;
```
Comment (sandreim) [inline src/arch/src/aarch64/regs.rs:77]: There seems to be a tab/spaces identation issue here.

File: src/arch/src/aarch64/regs.rs [hunk 2]
```diff
@@ -114,31 +122,47 @@ macro_rules! arm64_sys_reg {
...
+        PSTATE_FAULT_BITS_64,
+    )
+    .map_err(Error::SetCoreRegister)?;
 
     // Other vCPUs are powered off initially awaiting PSCI wakeup.
     if cpu_id == 0 {
         // Setting the PC (Processor Counter) to the current program address (kernel address).
-        vcpu.set_one_reg(arm64_core_reg!(pc), boot_ip)
+        let offset = offset__of!(user_pt_regs, pc) + kreg_off;
+        vcpu.set_one_reg(arm64_core_reg_id!(KVM_REG_SIZE_U64, offset), boot_ip as u64)
```
Comment (sandreim) [inline src/arch/src/aarch64/regs.rs:153]: The previous code was more readable, now I have a hard time figuring out the register which is on the line above. Can this be improved ?

File: src/arch/src/aarch64/regs.rs [hunk 3]
```diff
@@ -29,6 +27,15 @@ pub enum Error {
 }
 type Result<T> = result::Result<T, Error>;
 
+impl fmt::Debug for Error {
```
Comment (aghecenco) [inline src/arch/src/aarch64/regs.rs:30]: nit: this looks more like a `Display` impl, I think you could leave both the `Debug` derive (for unwraps / tests / ?) and convert this to a `Display`

File: src/arch/src/aarch64/regs.rs (hunk 2 above)
Comment (aghecenco) [reply src/arch/src/aarch64/regs.rs:153]: Maybe rename `offset` to a more verbose `pc_offset` (and respectively `pstate_offset` etc)?

File: src/arch/src/aarch64/regs.rs (hunk 3 above)
Comment (dianpopa) [reply src/arch/src/aarch64/regs.rs]: I applied your suggestion, take a look!

File: src/arch/src/aarch64/regs.rs (hunk 2 above)
Comment (dianpopa) [reply src/arch/src/aarch64/regs.rs]: @aghecenco I applied your suggestion, @sandreim take a look!

**Comment by dianpopa** [comment]
@aghecenco @sandreim I addressed your suggestions, please give it another look!

**Pull request merged**